classes/task/merge_user_accounts.php

```diff
@@ -0,0 +1,198 @@
+            $update
+        );
+    }
+    /**
+     * Function for verifying the users to keep and remove.
+     */
+    private function verify_users_to_keep_and_remove(object $record): void {
+        global $DB;
+        $keepuserfield = $record->keepuserfield;
+        $keepuservalue = $record->keepuservalue;
+        $removeuserfield = $record->removeuserfield;
+        $removeuservalue = $record->removeuservalue;
+        // Verify user to remove.
+        $usertoremove = $DB->get_records(merge_request::TABLE_USERS, [$removeuserfield => $removeuservalue]);
+        if (is_null($record->removeuserid)) {
+            $this->removeuserid = $this->find_user_id_or_fail($record->removeuserfield, $record->removeuservalue);
+        }
+        // Verify user to keep.
+        $usertokeep = $DB->get_records(merge_request::TABLE_USERS, [$keepuserfield => $keepuservalue]);
+        if (is_null($record->keepuserid)) {
+            $this->keepuserid = $this->find_user_id_or_fail($record->keepuserfield, $record->keepuservalue);
+        }
+    }
+    /**
+     * Function for updating number of retries.
+     */
+    private function update_retries_in_table(int $idrecord,
+                                             int $retries): void {
+        global $DB;
+        $DB->update_record(
+            merge_request::TABLE_MERGE_REQUEST,
+            (object)[
+                'id' => $idrecord,
+                'retries' => $retries,
+                'timecompleted' >= time(),
+            ]
+        );
+    }
+    /**
+     * Function for updating removeuserid.
+     */
+    private function update_removeuserid_in_table(int $idrecord,
+                                                        int $removeuserid): void {
+        global $DB;
+        $DB->update_record(
+            merge_request::TABLE_MERGE_REQUEST,
+            (object)[
+                'id' => $idrecord,
+                'removeuserid' => $removeuserid
+            ]
+        );
+    }
+    /**
+     * Function for updating keepuserid.
+     */
+    private function update_keepuserid_in_table(int $idrecord,
+                                                int $keepuserid) {
+        global $DB;
+        $DB->update_record(
+            merge_request::TABLE_MERGE_REQUEST,
+            (object)[
+                'id' => $idrecord,
+                'keepuserid' => $keepuserid
+            ]
+        );
+    }
+    /**
+     * Function to find user id or fail.
+     */
+    protected function find_user_id_or_fail(int $mergerequestid,
+                                            string $userfield, string $uservalue): int {
+        global $DB;
+        $users = $DB->get_records(merge_request::TABLE_USERS, [$userfield => $uservalue]);
+        if (count($users) == 0) {
+            throw new moodle_exception(get_string('cannotfinduser', 'tool_mergeusers',
+                                        (object)['userfield' => $userfield, 'uservalue' => $uservalue]));
```

Comment:
Any exception will make a proper log into the moodle execution task log, but not in the merge request logs. 

Exceptions from Moodle core while merging are captured by the MergeUserTool, and put inside the merge request logs in the prior plugin version. These exceptions that we are throwing here makes no log into the merge request and we are losing details from the process.

Then, I think it is necessary to log into the mergerequest. I would modify this method by passing the `mergerequest $mergerequest` parameter, instead of the current `int $mergerequestid`. Then, I will do these two tasks:

1. I will add the number of `$mergerequest->retries` into the `cannotfinduser` and `toomanyusers`.
2. I will add a log line, like `$mergerequest->appendlog([$errorfound], time());`, to record into the mergerequest why the current retry was failed.

To keep most of the code I would refactor this like:

```php
$errorfound = '';
if (count($users) == 0) {
     $errorfound = get_string(...., (object)['userfield'... 'uservalue'... 'retries']);
} else if (count($users) > 1) {
    $errorfound = ...;
}
if (!empty($errorfound)) {
     $mergerequest->appendlog([$errorfound], time());
     $this->update_log($mergerequest); // to store the whole set of logs into the database.
     throw new moodle_exception($errorfound);
}
```

Reply:
this is still pending